File: sdk/typescript/src/agent.ts
/**
 * Agent SDK client for AI agent subprocess code to communicate with AGEZT.
 *
 * This module provides an `AgentClient` that connects to the AGEZT Agent Gateway
 * via a Unix domain socket. The gateway accepts scoped JWT tokens that limit
 * what capabilities the agent subprocess can access.
 *
 * @example
 * ```ts
 * import { AgentClient } from "@agezt/sdk";
 *
 * const client = new AgentClient({
 *   token: process.env.AGEZT_AGENT_TOKEN!,
 * });
 *
 * // Remember a fact
 * const record = await client.memory.write({
 *   type: "fact",
 *   subject: "test",
 *   content: "This is a test",
 * });
 *
 * // Search memories
 * const results = await client.memory.search("test");
 * for (const r of results) {
 *   console.log(r.subject, r.content);
 * }
 *
 * // Publish an event
 * await client.eventbus.publish("my.event", { key: "value" });
 *
 * // Subscribe to events
 * for await (const event of client.eventbus.subscribe("my.>")) {
 *   console.log(event);
 * }
 * ```
 */

import * as http from "node:http";
import { ConfigAccessError } from "./errors.js";

/** Default socket path for the agent gateway. */
export const DEFAULT_SOCKET_PATH = "@agezt/agentgw.sock";

/** Capability constants for agent subprocess tokens. */
export const Capability = {
  // Eventbus capabilities
  EVENTBUS_PUBLISH: "eventbus.publish",
  EVENTBUS_SUBSCRIBE: "eventbus.subscribe",

  // Memory capabilities
  MEMORY_READ: "memory.read",
  MEMORY_WRITE: "memory.write",
  MEMORY_DELETE: "memory.delete",
  MEMORY_SEARCH: "memory.search",
  MEMORY_LIST: "memory.list",

  // Log capabilities
  LOG_READ: "log.read",
  LOG_WRITE: "log.write",

  // Agent capabilities
  AGENT_LIST: "agent.list",
  AGENT_QUERY: "agent.query",

  // Config capabilities
  CONFIG_ACCESS: "config.access",
  CONFIG_LIST: "config.list",
  CONFIG_SEARCH: "config.search",
} as const;

export type Capability = (typeof Capability)[keyof typeof Capability];

/** Error from the agent gateway. */
export class AgentError extends Error {
  constructor(
    public readonly code: string,
    public readonly message: string,
    public readonly statusCode: number = 500
  ) {
    super(`[${code}] ${message}`);
    this.name = "AgentError";
  }
}

/** Memory record structure. */
export interface MemoryRecord {
  id: string;
  type: string;
  subject: string;
  content: string;
  confidence: number;
  last_seen_ms: number;
  tags?: Record<string, string>;
}

/** Search result with relevance score. */
export interface SearchResult {
  id: string;
  type: string;
  subject: string;
  content: string;
  confidence: number;
  last_seen_ms: number;
  score: number;
}

/** Agent profile from the roster. */
export interface AgentProfile {
  id: string;
  slug: string;
  name: string;
  model: string;
  enabled: boolean;
  retired: boolean;
}

/** Event from the bus. */
export interface BusEvent {
  id: string;
  seq: number;
  ts_unix_ms: number;
  subject: string;
  actor: string;
  kind: string;
  correlation_id?: string;
  causation_id?: string;
  payload?: unknown;
  tags?: Record<string, string>;
}

/** Options for creating an AgentClient. */
export interface AgentClientOptions {
  /** JWT capability token from the parent agent. */
  token: string;
  /** Path to the agent gateway Unix socket. */
  socketPath?: string;
  /** Per-request timeout in milliseconds (default 30000). */
  timeoutMs?: number;
}

/**
 * Synchronous client for AI agent subprocess code to access AGEZT.
 *
 * Connects to the AGEZT Agent Gateway via a Unix domain socket and authenticates
 * using a scoped JWT token. The token is typically provided by the parent agent
 * via environment variable `AGEZT_AGENT_TOKEN`.
 */
export class AgentClient {
  private readonly token: string;
  private readonly socketPath: string;
  private readonly timeoutMs: number;

  /** Memory operations handle. */
  readonly memory: MemoryHandle;
  /** Eventbus operations handle. */
  readonly eventbus: EventbusHandle;
  /** Log operations handle. */
  readonly log: LogHandle;
  /** Agent operations handle. */
  readonly agent: AgentHandle;
  /** Config operations handle. */
  readonly config: ConfigHandle;

  constructor(opts: AgentClientOptions) {
    this.token = opts.token;
    this.socketPath = opts.socketPath ?? DEFAULT_SOCKET_PATH;
    this.timeoutMs = opts.timeoutMs ?? 30000;

    this.memory = new MemoryHandle(this);
    this.eventbus = new EventbusHandle(this);
    this.log = new LogHandle(this);
    this.agent = new AgentHandle(this);
    this.config = new ConfigHandle(this);
  }

  /** @internal Make a GET request to the gateway. */
  async get<T>(path: string): Promise<T> {
    return this.request<T>("GET", path);
  }

  /** @internal Make a POST request to the gateway. */
  async post<T>(path: string, body: Record<string, unknown>): Promise<T> {
    return this.request<T>("POST", path, body);
  }

  /** @internal Make a DELETE request to the gateway. */
  async delete<T>(path: string): Promise<T> {
    return this.request<T>("DELETE", path);
  }

  /** @internal Make a request to the gateway via Node.js HTTP-over-Unix socket. */
  private async request<T>(method: string, path: string, body?: Record<string, unknown>): Promise<T> {
    return new Promise((resolve, reject) => {
      const options: http.RequestOptions = {
        path,
        method,
        socketPath: this.socketPath,
        timeout: this.timeoutMs,
        headers: {
          Accept: "application/json",
          Authorization: `Bearer ${this.token}`,
          "Content-Type": "application/json",
        },
      };

      if (body !== undefined) {
        const bodyStr = JSON.stringify(body);
        (options.headers as Record<string, string>)["Content-Length"] = Buffer.byteLength(bodyStr).toString();
      }

      const req = http.request(options, (res) => {
        const chunks: Buffer[] = [];
        res.on("data", (chunk: Buffer) => chunks.push(chunk));
        res.on("end", () => {
          const respBody = Buffer.concat(chunks).toString();

          if (res.statusCode === 401) {
            reject(new AgentError("UNAUTHORIZED", "Invalid or missing token", 401));
            return;
          }
          if (res.statusCode === 403) {
            reject(new AgentError("FORBIDDEN", "Capability not granted", 403));
            return;
          }
          if (res.statusCode === 429) {
            reject(new AgentError("RATE_LIMITED", "Too many requests", 429));
            return;
          }
          if (res.statusCode && res.statusCode >= 400) {
            try {
              const err = JSON.parse(respBody);
              reject(new AgentError(err.error?.code ?? "ERROR", err.error?.message ?? respBody, res.statusCode));
            } catch {
              reject(new AgentError("HTTP_ERROR", respBody, res.statusCode ?? 500));
            }
            return;
          }

          try {
            resolve(respBody ? (JSON.parse(respBody) as T) : ({} as T));
          } catch {
            reject(new AgentError("INVALID_RESPONSE", "Cannot parse JSON response", 500));
          }
        });
      });

      req.on("error", (err) => {
        if ((err as NodeJS.ErrnoException).code === "ENOENT" || (err as NodeJS.ErrnoException).code === "ECONNREFUSED") {
          reject(new AgentError("CONNECTION_ERROR", `Cannot connect to gateway at ${this.socketPath}`, 503));
        } else {
          reject(new AgentError("REQUEST_ERROR", err.message, 500));
        }
      });

      req.on("timeout", () => {
        req.destroy();
        reject(new AgentError("TIMEOUT", `Request timed out after ${this.timeoutMs}ms`, 408));
      });

      if (body !== undefined) {
        req.write(JSON.stringify(body));
      }
      req.end();
    });
  }
}

/** Handle for memory operations. */
export class MemoryHandle {
  constructor(private client: AgentClient) {}

  /**
   * Remember (write) a memory record.
   *
   * @example
   * ```ts
   * const record = await client.memory.write({
   *   type: "fact",
   *   subject: "API endpoint",
   *   content: "Use POST /v1/memory/write",
   * });
   * console.log(record.id);
   * ```
   */
  async write(opts: {
    type?: string;
    subject?: string;
    content?: string;
    tags?: Record<string, string>;
  }): Promise<MemoryRecord> {
    const result = await this.client.post<{ record: MemoryRecord }>("/v1/memory/write", {
      type: opts.type ?? "fact",
      subject: opts.subject ?? "",
      content: opts.content ?? "",
      tags: opts.tags,
    });
    return result.record;
  }

  /**
   * Recall (search) memories matching a query.
   *
   * @example
   * ```ts
   * const results = await client.memory.search("API endpoint");
   * for (const r of results) {
   *   console.log(r.subject, r.score);
   * }
   * ```
   */
  async search(query: string, limit = 20): Promise<SearchResult[]> {
    const result = await this.client.get<{ results: SearchResult[] }>(
      `/v1/memory/search?q=${encodeURIComponent(query)}&limit=${limit}`
    );
    return result.results;
  }

  /**
   * Forget (delete) a specific memory record.
   *
   * @example
   * ```ts
   * const deleted = await client.memory.delete("01HABC...");
   * console.log(deleted);
   * ```
   */
  async delete(id: string): Promise<boolean> {
    const result = await this.client.delete<{ deleted: boolean }>(`/v1/memory/delete?id=${encodeURIComponent(id)}`);
    return result.deleted;
  }
}

/** Handle for eventbus operations. */
export class EventbusHandle {
  constructor(private client: AgentClient) {}

  /**
   * Publish an event to the bus.
   *
   * @example
   * ```ts
   * await client.eventbus.publish("agent.progress", {
   *   step: 1,
   *   total: 3,
   * });
   * ```
   */
  async publish(event: string, payload?: Record<string, unknown>): Promise<void> {
    await this.client.post("/v1/eventbus/publish", {
      event,
      payload,
    });
  }

  /**
   * Subscribe to events matching a pattern.
   *
   * @example
   * ```ts
   * for await (const ev of client.eventbus.subscribe("agent.>")) {
   *   console.log(ev.subject, ev.payload);
   * }
   * ```
   *
   * Note: This requires Node.js and the `node:http` module.
   * It uses Server-Sent Events (SSE) for streaming.
   */
  async *subscribe(pattern = ">"): AsyncGenerator<BusEvent> {
    const url = new URL(`http://localhost/v1/eventbus/subscribe?pattern=${encodeURIComponent(pattern)}`);

    const options: http.RequestOptions = {
      path: url.pathname + url.search,
      method: "GET",
      socketPath: (this.client as unknown as { socketPath: string }).socketPath,
      headers: {
        Accept: "text/event-stream",
        Authorization: `Bearer ${(this.client as unknown as { token: string }).token}`,
      },
    };

    const stream = await new Promise<ReadableStream<BusEvent>>((resolve, reject) => {
      const req = http.request(options, (res) => {
        if (!res.statusCode || res.statusCode >= 400) {
          reject(new AgentError("SUBSCRIBE_ERROR", `HTTP ${res.statusCode}`, res.statusCode ?? 500));
          return;
        }

        const stream = new ReadableStream<BusEvent>({
          start(controller) {
            // SSE parsing
            const decoder = new TextDecoder();
            let buffer = "";

            res.on("data", (chunk: Buffer) => {
              buffer += decoder.decode(chunk, { stream: true });

              // Process complete SSE messages
              const lines = buffer.split(/\r?\n/);
              buffer = lines.pop() ?? "";

              for (const line of lines) {
                if (line.startsWith("data:")) {
                  const data = line.slice(5).trim();
                  try {
                    const event = JSON.parse(data) as BusEvent;
                    controller.enqueue(event);
                  } catch {
                    // Skip invalid JSON
                  }
                }
              }
            });

            res.on("end", () => controller.close());
            res.on("error", (err) => controller.error(err));
          },
        });

        resolve(stream);
      });

      req.on("error", reject);
      req.on("timeout", () => {
        req.destroy();
        reject(new AgentError("TIMEOUT", "Subscription timed out", 408));
      });

      req.end();
    });

    const reader = stream.getReader();
    try {
      while (true) {
        const { done, value } = await reader.read();
        if (done) break;
        if (value) yield value;
      }
    } finally {
      reader.releaseLock();
    }
  }
}

/** Handle for log operations. */
export class LogHandle {
  constructor(private client: AgentClient) {}

  /**
   * Write a log entry.
   *
   * @example
   * ```ts
   * await client.log.write("Processing item", {
   *   level: "info",
   *   meta: { itemId: "123" },
   * });
   * ```
   */
  async write(message: string, opts?: { level?: string; meta?: Record<string, unknown> }): Promise<void> {
    await this.client.post("/v1/log/write", {
      level: opts?.level ?? "info",
      message,
      meta: opts?.meta,
    });
  }
}

/** Handle for agent operations. */
export class AgentHandle {
  constructor(private client: AgentClient) {}

  /**
   * List all agents in the roster.
   *
   * @example
   * ```ts
   * const agents = await client.agent.list();
   * for (const a of agents) {
   *   console.log(a.name, a.model);
   * }
   * ```
   */
  async list(): Promise<AgentProfile[]> {
    const result = await this.client.get<{ agents: AgentProfile[] }>("/v1/agent/list");
    return result.agents;
  }

  /**
   * Query a specific agent's status.
   *
   * @example
   * ```ts
   * const agent = await client.agent.query("claude");
   * console.log(agent.name, agent.enabled);
   * ```
   */
  async query(id: string): Promise<AgentProfile> {
    return this.client.get<AgentProfile>(`/v1/agent/query?id=${encodeURIComponent(id)}`);
  }
}

/** Handle for config operations. */
export class ConfigHandle {
  constructor(private client: AgentClient) {}

  /**
   * Get a config value by key.
   *
   * Config values are rated based on sensitivity:
   * - ``public``: Endpoint URLs, version info (auto-allowed)
   * - ``internal``: Application settings (auto-allowed)
   * - ``restricted``: Semi-sensitive values (requires HITL approval)
   * - ``secret``: API keys, tokens, passwords (always denied)
   *
   * @example
   * ```ts
   * // Public/internal values — automatic access
   * const endpoint = await client.config.get("analytics:endpoint");
   * const version = await client.config.get("app:version");
   *
   * // Secret values — always denied
   * try {
   *   const apiKey = await client.config.get("github:token");
   * } catch (err) {
   *   if (err instanceof ConfigAccessError) {
   *     console.log(`Access denied: ${err.code}`);
   *   }
   * }
   *
   * // Restricted values — may require HITL approval
   * try {
   *   const awsKey = await client.config.get(
   *     "aws:readonly_key",
   *     "Need AWS credentials to read S3 data"
   *   );
   * } catch (err) {
   *   if (err instanceof ConfigAccessError) {
   *     console.log(`Approval denied or timed out: ${err.message}`);
   *   }
   * }
   * ```
   *
   * @param key - Config key (e.g., "github:endpoint", "analytics:api_key")
   * @param reason - Why the agent needs this value (shown to operator for HITL approval)
   * @returns The config value
   * @throws {ConfigAccessError} When access is denied
   */
  async get(key: string, reason?: string): Promise<string> {
    let path = `/v1/config/${encodeURIComponent(key)}`;
    if (reason) {
      path += `?reason=${encodeURIComponent(reason)}`;
    }

    try {
      const result = await this.client.get<{ value: string }>(path);
      return result.value ?? "";
    } catch (err) {
      if (err instanceof ConfigAccessError) throw err;
      // Wrap other errors
      throw new ConfigAccessError(key, "INTERNAL_ERROR", String(err), 500);
    }
  }

  /**
   * List accessible config keys.
   *
   * Returns only keys with ``public`` or ``internal`` ratings.
   * ``secret`` and ``restricted`` keys are excluded.
   *
   * @example
   * ```ts
   * const keys = await client.config.listKeys();
   * for (const key of keys) {
   *   console.log(key);
   * }
   * ```
   */
  async listKeys(): Promise<string[]> {
    const result = await this.client.get<{ keys: string[] }>("/v1/config");
    return result.keys ?? [];
  }

  /**
   * Search config keys by prefix.
   *
   * Only ``public`` rating keys are searchable.
   *
   * @example
   * ```ts
   * const results = await client.config.search("github:");
   * for (const r of results) {
   *   console.log(r.key, r.description);
   * }
   * ```
   *
   * @param query - Search prefix (e.g., "github:", "analytics:")
   * @returns Matching entries with ``key`` and ``description``
   */
  async search(query: string): Promise<Array<{ key: string; description?: string }>> {
    const path = `/v1/config/search?q=${encodeURIComponent(query)}`;
    const result = await this.client.get<{ results: Array<{ key: string; description?: string }> }>(path);
    return result.results ?? [];
  }
}

File: sdk/python/agezt/agent.py
"""Agent SDK client for AI agent subprocess code to communicate with AGEZT.

This module provides an ``AgentClient`` that connects to the AGEZT Agent Gateway
via a Unix domain socket. The gateway accepts scoped JWT tokens that limit
what capabilities the agent subprocess can access.

Example usage::

    import os
    from agezt.agent import AgentClient

    # The token is passed via environment by the parent agent
    token = os.environ["AGEZT_AGENT_TOKEN"]
    client = AgentClient(token=token)

    # Remember a fact
    client.memory.write(type="fact", subject="test", content="This is a test")

    # Recall memories
    results = client.memory.search("test")
    for r in results:
        print(r["subject"], r["content"])

    # Publish an event
    client.eventbus.publish("my.event", {"key": "value"})

    # Subscribe to events
    for event in client.eventbus.subscribe("my.>"):
        print(event)

The client is thread-safe for basic use. For async usage, use ``AsyncAgentClient``.
"""

from __future__ import annotations

import json
import socket
import threading
from dataclasses import dataclass
from typing import Any, Dict, Iterator, List, Optional

__all__ = ["AgentClient", "AsyncAgentClient", "AgentError", "Capability"]

# Default socket path for the agent gateway
DEFAULT_SOCKET_PATH = "@agezt/agentgw.sock"


@dataclass
class Capability:
    """Represents a capability that can be granted to an agent subprocess."""

    # Eventbus capabilities
    EVENTBUS_PUBLISH = "eventbus.publish"
    EVENTBUS_SUBSCRIBE = "eventbus.subscribe"

    # Memory capabilities
    MEMORY_READ = "memory.read"
    MEMORY_WRITE = "memory.write"
    MEMORY_DELETE = "memory.delete"
    MEMORY_SEARCH = "memory.search"
    MEMORY_LIST = "memory.list"

    # Log capabilities
    LOG_READ = "log.read"
    LOG_WRITE = "log.write"

    # Agent capabilities
    AGENT_LIST = "agent.list"
    AGENT_QUERY = "agent.query"


class AgentError(Exception):
    """Raised when the agent gateway returns an error."""

    def __init__(self, code: str, message: str, status_code: int = 500) -> None:
        self.code = code
        self.message = message
        self.status_code = status_code
        super().__init__(f"[{code}] {message}")


class _SocketClient:
    """Low-level socket client for the agent gateway.
    
    Supports both Unix domain sockets and TCP sockets.
    Use socket_path in the form:
      - /path/to.sock for Unix domain socket
      - host:port for TCP socket (e.g., localhost:18765)
      - tcp://host:port for explicit TCP
    """

    def __init__(self, socket_path: str, timeout: float = 30.0) -> None:
        self.socket_path = socket_path
        self.timeout = timeout
        self._lock = threading.Lock()
        self._is_tcp = self._detect_tcp()

    def _detect_tcp(self) -> bool:
        """Detect if the socket path is a TCP address."""
        # If it starts with tcp:// it's TCP
        if self.socket_path.startswith("tcp://"):
            return True
        # If it looks like host:port (contains : but not at start), it's TCP
        if ":" in self.socket_path and not self.socket_path.startswith("/"):
            return True
        # Otherwise assume Unix socket
        return False

    def _create_socket(self) -> socket.socket:
        """Create the appropriate type of socket."""
        if self._is_tcp:
            sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        else:
            # Try Unix socket, fall back to TCP on Windows
            try:
                sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
            except (AttributeError, OSError):
                # AF_UNIX not available on Windows - use TCP loopback
                sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        return sock

    def _connect(self, sock: socket.socket) -> None:
        """Connect the socket to the gateway."""
        if self._is_tcp:
            if self.socket_path.startswith("tcp://"):
                addr = self.socket_path[6:]
            else:
                addr = self.socket_path
            host, port_str = addr.rsplit(":", 1)
            sock.connect((host, int(port_str)))
        else:
            sock.connect(self.socket_path)

    def _request(
        self,
        method: str,
        path: str,
        body: Optional[Dict[str, Any]] = None,
        headers: Optional[Dict[str, str]] = None,
    ) -> Dict[str, Any]:
        """Send a request and return the parsed JSON response."""
        with self._lock:
            sock = self._create_socket()
            sock.settimeout(self.timeout)
            try:
                self._connect(sock)
                try:
                    # Build HTTP request
                    req_lines = [f"{method} {path} HTTP/1.1"]
                    req_lines.append("Host: localhost")
                    req_lines.append("Accept: application/json")

                    if headers:
                        for k, v in headers.items():
                            req_lines.append(f"{k}: {v}")

                    if body is not None:
                        body_bytes = json.dumps(body).encode("utf-8")
                        req_lines.append(f"Content-Type: application/json")
                        req_lines.append(f"Content-Length: {len(body_bytes)}")
                    else:
                        body_bytes = b""

                    req_lines.append("")
                    req_lines.append("")

                    sock.sendall("\r\n".join(req_lines).encode("utf-8"))
                    if body_bytes:
                        sock.sendall(body_bytes)

                    # Read response
                    response = b""
                    while True:
                        chunk = sock.recv(4096)
                        if not chunk:
                            break
                        response += chunk

                    return self._parse_http_response(response)
                finally:
                    sock.close()
            except socket.timeout as e:
                raise AgentError("TIMEOUT", f"Request timed out after {self.timeout}s", 408) from e
            except OSError as e:
                raise AgentError("CONNECTION_ERROR", f"Cannot connect to gateway: {e}", 503) from e

    def _parse_http_response(self, response: bytes) -> Dict[str, Any]:
        """Parse HTTP response and return JSON body."""
        # Split headers and body
        parts = response.split(b"\r\n\r\n", 1)
        if len(parts) < 2:
            raise AgentError("INVALID_RESPONSE", "Invalid HTTP response from gateway", 500)

        headers = parts[0].decode("utf-8", errors="replace")
        body = parts[1]

        # Parse status line
        status_line = headers.split("\r\n")[0]
        try:
            status_code = int(status_line.split()[1])
        except (IndexError, ValueError):
            raise AgentError("INVALID_RESPONSE", f"Cannot parse status: {status_line}", 500)

        # Check for chunked encoding
        if "chunked" in headers.lower():
            body = self._decode_chunked(body)

        # Remove trailing \r\n
        body = body.rstrip(b"\r\n")

        if not body:
            return {}

        try:
            return json.loads(body.decode("utf-8"))
        except json.JSONDecodeError as e:
            raise AgentError("INVALID_RESPONSE", f"Cannot parse JSON: {e}", 500) from e

    def _decode_chunked(self, body: bytes) -> bytes:
        """Decode chunked transfer encoding."""
        result = b""
        pos = 0
        while pos < len(body):
            # Read chunk size line
            line_end = body.find(b"\r\n", pos)
            if line_end == -1:
                break
            chunk_size = int(body[pos:line_end].split(b";")[0], 16)
            if chunk_size == 0:
                break
            pos = line_end + 2
            result += body[pos : pos + chunk_size]
            pos += chunk_size + 2  # Skip \r\n after chunk
        return result


class _EventbusHandle:
    """Handle for eventbus operations."""

    def __init__(self, client: "_AgentClient") -> None:
        self._client = client

    def publish(self, event: str, payload: Optional[Dict[str, Any]] = None) -> None:
        """Publish an event to the bus.

        Args:
            event: The event type (subject) for routing.
            payload: Optional JSON-serializable payload.
        """
        body: Dict[str, Any] = {"event": event}
        if payload is not None:
            body["payload"] = payload

        self._client._post("/v1/eventbus/publish", body)

    def subscribe(self, pattern: str = ">") -> Iterator[Dict[str, Any]]:
        """Subscribe to events matching a pattern.

        Uses SSE (Server-Sent Events) for streaming.

        Args:
            pattern: Event pattern to match. Use ">" for all events.

        Yields:
            Dict: Each event as a dictionary.

        Example::

            for ev in client.eventbus.subscribe("my.>"):
                print(ev["subject"], ev["payload"])
        """
        return self._client._subscribe("/v1/eventbus/subscribe?pattern=" + pattern)


class _MemoryHandle:
    """Handle for memory operations."""

    def __init__(self, client: "_AgentClient") -> None:
        self._client = client

    def write(
        self,
        type: str = "fact",
        subject: str = "",
        content: str = "",
        tags: Optional[Dict[str, str]] = None,
    ) -> Dict[str, Any]:
        """Remember (write) a memory record.

        Args:
            type: Memory type (e.g., "fact", "idea", "todo", "learned").
            subject: Brief subject/title for the memory.
            content: The actual memory content.
            tags: Optional key-value tags.

        Returns:
            Dict: The created memory record with ID.
        """
        body: Dict[str, Any] = {
            "type": type,
            "subject": subject,
            "content": content,
        }
        if tags:
            body["tags"] = tags

        result = self._client._post("/v1/memory/write", body)
        return result.get("record", {})

    def search(self, query: str, limit: int = 20) -> List[Dict[str, Any]]:
        """Recall (search) memories matching a query.

        Args:
            query: Natural language search query.
            limit: Maximum number of results.

        Returns:
            List[Dict]: Matching memory records with relevance scores.
        """
        result = self._client._get(f"/v1/memory/search?q={query}&limit={limit}")
        return result.get("results", [])

    def delete(self, id: str) -> bool:
        """Forget (delete) a specific memory record.

        Args:
            id: The memory record ID to delete.

        Returns:
            bool: True if deleted, False if not found.
        """
        result = self._client._delete(f"/v1/memory/delete?id={id}")
        return result.get("deleted", False)


class _LogHandle:
    """Handle for logging operations."""

    def __init__(self, client: "_AgentClient") -> None:
        self._client = client

    def write(
        self, message: str, level: str = "info", meta: Optional[Dict[str, Any]] = None
    ) -> None:
        """Write a log entry.

        Args:
            message: The log message.
            level: Log level (debug, info, warn, error). Default: info.
            meta: Optional metadata dictionary.
        """
        body: Dict[str, Any] = {
            "level": level,
            "message": message,
        }
        if meta:
            body["meta"] = meta

        self._client._post("/v1/log/write", body)


class _AgentHandle:
    """Handle for agent operations."""

    def __init__(self, client: "_AgentClient") -> None:
        self._client = client

    def list(self) -> List[Dict[str, Any]]:
        """List all agents in the roster.

        Returns:
            List[Dict]: Agent profiles.
        """
        result = self._client._get("/v1/agent/list")
        return result.get("agents", [])

    def query(self, id: str) -> Dict[str, Any]:
        """Query a specific agent's status.

        Args:
            id: The agent ID or slug to query.

        Returns:
            Dict: Agent profile.
        """
        result = self._client._get(f"/v1/agent/query?id={id}")
        return result


class _ConfigHandle:
    """Access to the Config Center.

    Config values are rated based on their sensitivity:
    - ``public``: Endpoint URLs, version info (auto-allowed)
    - ``internal``: Application settings (auto-allowed)
    - ``restricted``: Semi-sensitive values (requires HITL approval)
    - ``secret``: API keys, tokens, passwords (always denied)

    Example::

        client = AgentClient(token=os.environ["AGEZT_AGENT_TOKEN"])

        # Public/internal values — automatic access
        endpoint = client.config.get("analytics:endpoint")
        version = client.config.get("app:version")

        # Secret values — always denied
        try:
            api_key = client.config.get("github:token")  # raises ConfigAccessError
        except ConfigAccessError as e:
            print(f"Access denied: {e.code}")

        # Restricted values — may require HITL approval
        try:
            aws_key = client.config.get(
                "aws:readonly_key",
                reason="Need AWS credentials to read S3 data"
            )
        except ConfigAccessError as e:
            print(f"Approval denied or timed out: {e.message}")
    """

    def __init__(self, client: "_AgentClient") -> None:
        self._client = client

    def get(self, key: str, reason: str = "") -> str:
        """Get a config value by key.

        Args:
            key: Config key (e.g., "github:endpoint", "analytics:api_key").
            reason: Why the agent needs this value. Shown to operator for
                HITL approval requests.

        Returns:
            str: The config value.

        Raises:
            ConfigAccessError: When access is denied. Possible codes:
                - ``ACCESS_DENIED``: Rating=secret or HITL denied/timeout
                - ``KEY_NOT_FOUND``: Config key does not exist
                - ``RATE_LIMITED``: Too many config requests
        """
        from agezt.errors import ConfigAccessError

        path = f"/v1/config/{key}"
        if reason:
            # URL-encode the reason
            import urllib.parse

            path += f"?{urllib.parse.urlencode({'reason': reason})}"

        try:
            result = self._client._get(path)
            return result.get("value", "")
        except Exception as e:
            # Re-raise as ConfigAccessError if it's an API error
            if isinstance(e, ConfigAccessError):
                raise
            # For other errors (network, etc.), wrap them
            raise ConfigAccessError(
                key=key,
                code="INTERNAL_ERROR",
                message=str(e),
                status=500,
            )

    def list_keys(self) -> List[str]:
        """List accessible config keys.

        Returns only keys with ``public`` or ``internal`` ratings.
        ``secret`` and ``restricted`` keys are excluded.

        Returns:
            List[str]: List of accessible config keys.
        """
        result = self._client._get("/v1/config")
        return result.get("keys", [])

    def search(self, query: str) -> List[Dict[str, Any]]:
        """Search config keys by prefix.

        Only ``public`` rating keys are searchable.

        Args:
            query: Search prefix (e.g., "github:", "analytics:").

        Returns:
            List[Dict]: Matching entries with ``key`` and ``description``.
        """
        import urllib.parse

        path = f"/v1/config/search?{urllib.parse.urlencode({'q': query})}"
        result = self._client._get(path)
        return result.get("results", [])


class _AgentClient:
    """Base class for agent client - handles the HTTP-over-Unix transport."""

    def __init__(
        self,
        token: str,
        socket_path: str = DEFAULT_SOCKET_PATH,
        timeout: float = 30.0,
    ) -> None:
        self.token = token
        self.socket_path = socket_path
        self.timeout = timeout
        self._sock = _SocketClient(socket_path, timeout)

        # Sub-handles
        self.eventbus = _EventbusHandle(self)
        self.memory = _MemoryHandle(self)
        self.log = _LogHandle(self)
        self.agent = _AgentHandle(self)
        self.config = _ConfigHandle(self)

    def _get(self, path: str) -> Dict[str, Any]:
        """Send GET request."""
        return self._sock._request(
            "GET",
            path,
            headers={"Authorization": f"Bearer {self.token}"},
        )

    def _post(self, path: str, body: Dict[str, Any]) -> Dict[str, Any]:
        """Send POST request."""
        return self._sock._request(
            "POST",
            path,
            body=body,
            headers={"Authorization": f"Bearer {self.token}"},
        )

    def _delete(self, path: str) -> Dict[str, Any]:
        """Send DELETE request."""
        return self._sock._request(
            "DELETE",
            path,
            headers={"Authorization": f"Bearer {self.token}"},
        )

    def _subscribe(self, path: str) -> Iterator[Dict[str, Any]]:
        """Subscribe to SSE stream."""
        # SSE subscription requires special handling
        sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        sock.settimeout(self.timeout)
        sock.connect(self.socket_path)

        try:
            # Build HTTP request for SSE
            req = (
                f"GET {path} HTTP/1.1\r\n"
                f"Host: localhost\r\n"
                f"Accept: text/event-stream\r\n"
                f"Authorization: Bearer {self.token}\r\n"
                f"Cache-Control: no-cache\r\n"
                f"\r\n"
            )
            sock.sendall(req.encode("utf-8"))

            # Read SSE stream
            buffer = b""
            while True:
                chunk = sock.recv(4096)
                if not chunk:
                    break
                buffer += chunk

                # Process SSE events
                while b"\r\n\r\n" in buffer:
                    header_end = buffer.index(b"\r\n\r\n")
                    headers = buffer[:header_end].decode("utf-8", errors="replace")
                    buffer = buffer[header_end + 4 :]

                    # Look for SSE data
                    for line in headers.split("\r\n"):
                        if line.startswith("data:"):
                            data = line[5:].strip()
                            if data:
                                try:
                                    yield json.loads(data)
                                except json.JSONDecodeError:
                                    pass
        finally:
            sock.close()


class AgentClient(_AgentClient):
    """Synchronous client for AI agent subprocess code to access AGEZT.

    Connects to the AGEZT Agent Gateway via a Unix domain socket and authenticates
    using a scoped JWT token. The token is typically provided by the parent agent
    via environment variable ``AGEZT_AGENT_TOKEN``.

    Args:
        token: JWT capability token from the parent agent.
        socket_path: Path to the agent gateway Unix socket.
        timeout: Request timeout in seconds.

    Example::

        import os
        from agezt import AgentClient

        client = AgentClient(token=os.environ["AGEZT_AGENT_TOKEN"])

        # Remember
        client.memory.write(type="fact", subject="API", content="Use POST /v1/memory/write")

        # Search
        results = client.memory.search("API")
        for r in results:
            print(r["subject"], r["content"])

        # Publish event
        client.eventbus.publish("agent.progress", {"step": 1, "total": 3})
    """

    pass


# Note: AsyncAgentClient would require aiohttp or similar for proper async socket support.
# For now, the sync AgentClient works well with threading in async frameworks like asyncio.
# A future version can add async support when needed.
